src/components/assetlib_projects/asset_library_projects.gd

```diff
@@ -0,0 +1,353 @@
+		if not value:
+			_search_field.editable = false
+			_version_option.disabled = true
+			_sort_option.disabled = true
+			_category_option.disabled = true
+			_site_option.disabled = true
+		else:
+			_search_field.editable = true
+			_version_option.disabled = false
+			_sort_option.disabled = false
+			_category_option.disabled = false
+			_site_option.disabled = false
+
+@onready var _search_field: LineEdit = %SearchField
+@onready var _version_option: OptionButton = %VersionOption
+@onready var _sort_option: OptionButton = %SortOption
+@onready var _category_option: OptionButton = %CategoryOption
+@onready var _site_option: OptionButton = %SiteOption
+@onready var _support_options: MenuButton = %SupportOptions
+@onready var _asset_querier: HTTPRequest = $AssetQuerier
+@onready var _asset_list: HFlowContainer = %AssetList
+@onready var _navigation_buttons: HBoxContainer = %NavigationButtons
+@onready var _page_buttons: HBoxContainer = _navigation_buttons.get_node("PageButtons")
+@onready var _first_button: Button = _navigation_buttons.get_node("FirstButton")
+@onready var _previous_button: Button = _navigation_buttons.get_node("PreviousButton")
+@onready var _next_button: Button = _navigation_buttons.get_node("NextButton")
+@onready var _last_button: Button = _navigation_buttons.get_node("LastButton")
+@onready var _color_rect: ColorRect = %ColorRect
+@onready var _status_label: Label = %StatusLabel
+@onready var _refresh_button: Button = %RefreshButton
+@onready var _overlay_contents: CenterContainer = %OverlayContents
+
+
+func _ready():
+	$MarginContainer/ScrollContainer.add_theme_stylebox_override("panel",
+			get_theme_stylebox("search_panel", "ProjectManager"))
+	
+	_fetched_versions = false
+	
+	_support_options.get_popup().id_pressed.connect(_fetch_assets)
+	
+	# Populate PageButtons
+	for i in _MAX_PAGE_BUTTONS:
+		var button = Button.new()
+		button.name = "PageButton" + str(i)
+		_page_buttons.add_child(button)
+		button.pressed.connect(func():
+				_on_page_button_pressed(button)
+		)
+	
+	_first_button.pressed.connect(func():
+		_current_page = 0
+	)
+	_previous_button.pressed.connect(func():
+		_current_page = max(0, _current_page - 1)
+	)
+	_next_button.pressed.connect(func():
+		if _current_assets:
+			_current_page = min(_current_assets.pages - 1, _current_page + 1)
+		else:
+			_current_page = 0
+	)
+	_last_button.pressed.connect(func():
+		_current_page = _current_assets.pages - 1 if _current_assets else 0
+	)
+	
+	await _setup_version_button()
+	
+	_fetch_assets()
+
+
+func _setup_version_button():
+	var versions = await _fetch_versions()
+	versions.reverse()
+	if versions == []:
+		_version_option.disabled = true
+		return
+	
+	for version in versions:
+		_version_option.add_item(version)
+
+
+func _on_search_field_text_changed(_new_text: String):
+	_last_text_edit = Time.get_ticks_msec()
+	await get_tree().create_timer(1).timeout
+	if Time.get_ticks_msec() - _last_text_edit >= 1000:
+		_fetch_assets()
```

Comment:
We are waiting for the user to stop typing for longer than one second to send the request, because I feared we'd be DOSing the asset library if we were to send a request every time the user types something.

Reply:
await is pretty messy thing i would like to get rid of (because it should be handled like in ::113 with some extra checks in order to prevent things be called twice)

use debounce timer for doing so:

```gdscript
var debounce_time = 0.25

debounce_timer.one_shot = true
debounce_timer.timeout.connect(func(): _fetch_assets())

func _on_search_field_text_changed(_new_text: String):
    debounce_timer.start(debounce_time)
```

Reply:
@MakovWait said:
> ```gdscript
> var debounce_time = 0.25
> ```

If we consider the average human typing speed to be 40 wpm (which, if we assume one English word to be 5 characters long, would be equivalent to 200 cpm), then each character would take on average 0.3 seconds to be typed. Maybe we should use a value higher than 0.25 s?

Reply:
oh yea, it is just an example.